GlobalDescriptorTable: option/feature to increase number of entries

Hi there!

Would it be possible to increase the number of entries in the GDT? If you'd prefer to keep the default at 8, perhaps there could be a feature to increase it (to 16, perhaps)? To be able to use `syscall`/`sysret`, I reckon you need at least 7 entries, not including the null first entry. 2 for kernel code/data, 1 for the TSS, 2 for 32-bit user code/data and 2 for 64-bit code/data. If you also want an entry to allow CPU-local data with GS, then the current type panics. It's entirely possible I'm wrong, which would be great.

Thanks for all your hard work.

> [`feature(const_generics_defaults)`](https://github.com/rust-lang/rust/pull/90207) would make it possible to keep the current maximum (8) by default, which would also make this a non-breaking change. 

Unfortunately, this doesn't work, I misunderstood how much Rust can infer automatically. 
`feature(const_generics_defaults)` makes this work:
```rust
let mut gdt: GlobalDescriptorTable = GlobalDescriptorTable::new();
```
However this does not work:
```rust
let mut gdt = GlobalDescriptorTable::new();
```
```
error[E0282]: type annotations needed for `GlobalDescriptorTable<{_: usize}>`
 --> src/structures/gdt.rs:39:15
  |
6 | let mut gdt = GlobalDescriptorTable::new();
  |     -------   ^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer the value of const parameter `MAX_CAPACITY`
  |     |
  |     consider giving `gdt` the explicit type `GlobalDescriptorTable<MAX_CAPACITY>`, where the const parameter `MAX_CAPACITY` is specified
```
We could work around this by always returning `GlobalDescriptorTable<8>` from `GlobalDescriptorTable::new` and add a different constructor that allows variable sizes until the next breaking change, but this feel a bit too hackish IMO.

---

To be honest, I'm not so sure anymore if const generics are the right way to solve this: A lot of our users will probably never need anything but the default, so adding const generics would add unnecessary complexity for most. I thought we could avoid this using default values, but as I already said, this turned out to be incorrect.

Simply increasing the capacity will work for now, but only until the next person requires a bigger capacity. This might still be the most pragmatic solution though.

Another possible solution might be to allow creating GDTs of dynamic lengths at runtime by supplying a slice of `u64`s as backing memory. This would probably need another type though, I don't see how we could integrate this into the existing GDT type.

I'm open to any other suggestions or feedback!

--- a/src/structures/gdt.rs
+++ b/src/structures/gdt.rs
@@ -75,7 +75,7 @@ impl<const MAX: usize> GlobalDescriptorTable<MAX> {
     /// # Safety
     ///
     /// * The user must make sure that the entries are well formed
-    /// * The provided slice **must not be larger than 8 items** (only up to the first 8 will be observed.)
+    /// * Panics if the provided slice has more than `MAX` entries
     #[inline]
     pub const unsafe fn from_raw_slice(slice: &[u64]) -> Self {
         let next_free = slice.len();
@@ -84,7 +84,7 @@ impl<const MAX: usize> GlobalDescriptorTable<MAX> {
 
         assert!(
             next_free <= MAX,
-            "initializing a GDT from a slice requires it to be **at most** 8 elements."
+            "cannot initialize GDT with slice exceeding the maximum length"
         );
 
         while idx != next_free {

--- a/src/structures/gdt.rs
+++ b/src/structures/gdt.rs
@@ -14,10 +14,13 @@ use crate::registers::segmentation::{Segment, CS, SS};
 /// In 64-bit mode, segmentation is not supported. The GDT is used nonetheless, for example for
 /// switching between user and kernel mode or for loading a TSS.
 ///
-/// The GDT has a fixed size of 8 entries, trying to add more entries will panic.
+/// The GDT has a fixed maximum size given by the `MAX` const generic parameter.
+/// Trying to add more entries than this maximum via [`GlobalDescriptorTable::add_entry`]
+/// will panic.
 ///
 /// You do **not** need to add a null segment descriptor yourself - this is already done
-/// internally.
+/// internally. This means you can add up to `MAX - 1` additional [`Descriptor`]s to
+/// this table.
 ///
 /// Data segment registers in ring 0 can be loaded with the null segment selector. When running in
 /// ring 3, the `ss` register must point to a valid data segment which can be obtained through the
@@ -45,17 +48,24 @@ use crate::registers::segmentation::{Segment, CS, SS};
 /// ```
 
 #[derive(Debug, Clone)]
-pub struct GlobalDescriptorTable {
-    table: [u64; 8],
+pub struct GlobalDescriptorTable<const MAX: usize = 8> {
+    table: [u64; MAX],
     next_free: usize,
 }
 
 impl GlobalDescriptorTable {
-    /// Creates an empty GDT.
+    /// Creates an empty GDT with the default length of 8.
+    pub const fn new() -> Self {
+        Self::empty()
+    }
+}
+
+impl<const MAX: usize> GlobalDescriptorTable<MAX> {
+    /// Creates an empty GDT which can hold `MAX` number of [`Descriptor`]s.
     #[inline]
-    pub const fn new() -> GlobalDescriptorTable {
-        GlobalDescriptorTable {
-            table: [0; 8],
+    pub const fn empty() -> Self {
+        Self {
+            table: [0; MAX],
             next_free: 1,
         }
     }
@@ -67,13 +77,13 @@ impl GlobalDescriptorTable {
     /// * The user must make sure that the entries are well formed
     /// * The provided slice **must not be larger than 8 items** (only up to the first 8 will be observed.)
     #[inline]
-    pub const unsafe fn from_raw_slice(slice: &[u64]) -> GlobalDescriptorTable {
+    pub const unsafe fn from_raw_slice(slice: &[u64]) -> Self {
         let next_free = slice.len();
-        let mut table = [0; 8];
+        let mut table = [0; MAX];
         let mut idx = 0;
 
         assert!(
-            next_free <= 8,
+            next_free <= MAX,
             "initializing a GDT from a slice requires it to be **at most** 8 elements."
         );
 
@@ -82,7 +92,7 @@ impl GlobalDescriptorTable {
             idx += 1;
         }
 
-        GlobalDescriptorTable { table, next_free }
+        Self { table, next_free }
     }
 
     /// Get a reference to the internal table.

